### DISCUSSION TITLE
关于crash的问题

### DISCUSSION THREAD

**Discussion by novada0** [SUSTech-CS305-Fall22/CS305-Project-Skeleton#72] [Q&A]
1、crash_test引起crash的原因好像是在终端终止peer的，我们可以在主循环检测到KeyboardInterrupt的时候就找新的peer吗？而不是在等到10s或者15s之后才找。
2、不太理解为什么发的是WHOHAS，如果有记录某个chunk哪些peer有，是不是直接找到另一个peer发送GET就可以了呢？
3、如果第二个问题可以发送GET的话，应该如何做到concurrency呢？因为有可能会出现：两个peer向同一个peer请求同一个chunk的情况（如果header标识chunkhash的话会出问题），或者一个peer向同一个peer请求两个chunk的情况（只能在header中标识chunkhash）。
谢谢！！！

**Comment by TANSixu** [comment]
Hi，
1. KeyboardInterrupt只会发送给被关闭的peer，其他peer进程不会收到这个信号
2. 发WHOHAS是最简单的设计，相当于直接重新启动这个任务。实现记录哪些peer有然后直接发GET当然是很棒的设计，是绝对可以的！但是如果现在还没有实现这里的同学可以直接用最简单的方案，把注意力更多放在congestion control和concurrency上，因为comprehensive test已经移除了crash
3. a. 两个peer像统一额peer请求同一个chunk可以在发送端用地址来区分不同连接。
3. b. 如文档提到，一个发送端不会向同一个接收端混发两个chunk的包，所以在这个情况下接收端会先下载完第一个chunk再下载第二个。对于同一个发送端来说，一个接收端只会有一个正在下载的chunk。
